cmake/match.py

```diff
@@ -5,39 +5,137 @@
+    print("Line " + str(match_line) + " does not match")
+    print("is:       " + present)
+    print("expected: " + expected)
+
+
+## @brief pattern matching script status values
+class Status(Enum):
+    INPUT_END = 1
+    MATCH_END = 2
+    INPUT_AND_MATCH_END = 3
+    PROCESSING = 4
+
+
+## @brief check matching script status
+def check_status(input_lines, match_lines):
+    if not input_lines and not match_lines:
+        return Status.INPUT_AND_MATCH_END
+    elif not input_lines:
+        return Status.INPUT_END
+    elif not match_lines:
+        return Status.MATCH_END
+    return Status.PROCESSING
 
-for i, match_line in enumerate(match_lines):
-    # split into parts at {{ }}
-    match_parts = re.split(r'\{{(.*?)\}}', match_line.strip())
-    pattern = ""
-    for j, part in enumerate(match_parts):
-        if j % 2 == 0:
-            pattern += re.escape(part)
-        else:
-            pattern += part
 
-    input_line = input_lines[i].strip()
-    if not re.fullmatch(pattern, input_line):
-        print(f"Line {i+1} does not match".format(i+1))
-        print("is:       " + input_line)
-        print("expected: " + match_line.strip())
+## @brief pattern matching tags.
+## Tags are expected to be at the start of the line.
+class Tag(Enum):
+    OPT = "{{OPT}}"         # makes the line optional
+    IGNORE = "{{IGNORE}}"   # ignores all input until next match or end of input file
+
+
+## @brief main function for the match file processing script
+def main():
+    if len(sys.argv) != 3:
+        print("Usage: python match.py <input_file> <match_file>")
         sys.exit(1)
+
+    input_file = sys.argv[1]
+    match_file = sys.argv[2]
+
+    with open(input_file, 'r') as input, open(match_file, 'r') as match:
+        input_lines = input.readlines()
+        match_lines = match.readlines()
+
+    ignored_lines = []
+
+    input_idx = 0
+    match_idx = 0
+    tags_in_effect = []
+    status = Status.PROCESSING
+    while status == Status.PROCESSING:
+        status = check_status(input_lines[input_idx:], match_lines[match_idx:])
+        if status == Status.INPUT_AND_MATCH_END:
+            sys.exit(0)
+        elif status == Status.MATCH_END:
+            # last line of the match file is {{IGNORE}} tag, stop processing
+            if Tag.IGNORE in tags_in_effect:
+                sys.exit(0)
+            else:
+                print_incorrect_match(match_idx + 1, input_lines[input_idx].strip(), "");
+                print_content(input_lines, match_lines, ignored_lines)
+                sys.exit(1)
+        elif status == Status.INPUT_END:
+            print_incorrect_match(match_idx + 1, "", match_lines[match_idx].strip());
+            print_content(input_lines, match_lines, ignored_lines)
+            sys.exit(1)
+        else:
+            input_line = input_lines[input_idx].strip()
+
+        match_line = match_lines[match_idx]
+        if match_line.startswith(Tag.OPT.value):
+            tags_in_effect.append(Tag.OPT)
```

Comment:
doesn't seem like there can be more than one active tag?

Reply:
In the case below, there would be a point when ignore tag (active until the next match) is applied, then the optional tag (active until the next line) is applied next.
```
{{IGNORE}}
{{OPT}} optional line
```